debian/docker.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH DOCKER 1 "November 2, 2002"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
docker \- System tray for KDE3/GNOME2 docklet applications 
.SH SYNOPSIS
.B docker
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B docker
command.
This manual page was written for the Debian distribution
because the original program does not have a manual page.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBdocker\fP is a docking application (WindowMaker dock app) which acts as a system tray for KDE3 and GNOME2. It can be used to replace the panel in either
environment, allowing you to have a system tray without running the KDE/GNOME
panel.
.PP
Docker was written and designed to work with Openbox 2, but it should work
fine in any window manager that supports WindowMaker dock apps.

.SH OPTIONS
.TP
.B \-h-help
Show summary of options.
.TP
.B \-display DISPLAY
The X display to connect to.
.TP
.B \-border N
The width of the border to put around the system tray icons. Defaults to 1.
.TP
.B \-vertical
Line up the icons vertically. Defaults to horizontally.
.TP
.B \-wmaker
WindowMaker mode. This makes docker a fixed size (64x64) to appear nicely in in
WindowMaker.  Note: In this mode, you have a fixed number of icons that docker
can hold.
.TP
.B \-iconsize SIZE
The size (width and height and height) to display icons as in the system
tray. Defaults to 24.
.TP
.B \-color COLOR
The background color to use for the tray. Defaults to whatever color the window
manager specifies.
.SH AUTHOR
This manual page was written by Ari Pollak <ari@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
docker (1.5-1) UNRELEASED; urgency=medium

  * New upstream release (Closes: #292486)
  * Change installed binary name to "wmdocker" (Closes: #740863)

 -- Tianon Gravi <admwiggin@gmail.com>  Mon, 21 Apr 2014 09:34:40 -0600

docker (1.4-5) unstable; urgency=low

  * Convert patches to quilt
  * 02_no-poll.patch:
    - Apply patch from Leandro Lucarella to stop polling every 1/2 second
      (Closes: #501094)
  * Move menu entry to Applications/Viewers (Closes: #482928)

 -- Ari Pollak <ari@debian.org>  Sat, 25 Oct 2008 15:35:44 -0400

docker (1.4-4) unstable; urgency=low

  * Add -color option to allow setting the default background color with
    standard X color specifications (Closes: #416240)
  * Fix confusing package description (Closes: #450751)
  * Fix spelling error in help text (Closes: #450749)
  * Fix broken watch file (Closes: #449669)
  * Fix misleading Info reference in manpage (Closes: #295873)
  * Add -wmaker option to manpage (Closes: #326777)

 -- Ari Pollak <ari@debian.org>  Sat, 01 Dec 2007 00:43:47 -0500

docker (1.4-3) unstable; urgency=low

  * Depend on libx11-dev | xlibs-dev instead of xlibs-dev

 -- Ari Pollak <ari@debian.org>  Mon,  5 Dec 2005 10:47:35 -0500

docker (1.4-2) unstable; urgency=low

  * Check for noopt in DEB_BUILD_OPTIONS correctly

 -- Ari Pollak <ari@debian.org>  Thu, 14 Nov 2002 19:10:14 -0500

docker (1.4-1) unstable; urgency=low

  * New upstream release
    - Aesthetic support for WindowMaker has been added with the -wmaker option 

 -- Ari Pollak <ari@debian.org>  Sun, 10 Nov 2002 23:14:21 -0500

docker (1.3-1) unstable; urgency=low

  * Initial Release. (Closes: #167491)

 -- Ari Pollak <ari@debian.org>  Sat,  2 Nov 2002 15:11:22 -0500



docker.c
#include "version.h"
#include "kde.h"
#include "icons.h"
#include "docker.h"
#include "net.h"
/*#include <X11/Xlib.h>*/
#include <X11/Xutil.h>
/*#include <X11/Xatom.h>
#include <glib.h>*/
#include <assert.h>
#include <signal.h>
#include <stdlib.h>

int argc;
char **argv;

Window win = None, hint_win = None, root = None;
gboolean wmaker = FALSE; /* WindowMakerMode!!! wheeee */
Display *display = NULL;
GSList *icons = NULL;
int width = 0, height = 0;
int border = 1; /* blank area around icons. must be > 0 */
gboolean horizontal = TRUE; /* layout direction */
int icon_size = 24; /* width and height of systray icons */

static char *display_string = NULL;
/* excluding the border. sum of all child apps */
static gboolean exit_app = FALSE;

void signal_handler(int signal)
{
  switch (signal)
  {
  case SIGSEGV:
  case SIGFPE:
    g_printerr("Segmentation Fault or Critical Error encountered. "
               "Dumping core and aborting.\n");
    abort();
    
  case SIGPIPE:
  case SIGTERM:
  case SIGINT:
  case SIGHUP:
    exit_app = TRUE;
  }
}


void parse_cmd_line()
{
  int i;
  gboolean help = FALSE;
  
  for (i = 1; i < argc; i++) {
    if (0 == strcasecmp(argv[i], "-display")) {
      ++i;
      if (i < argc) {
        display_string = argv[i];
      } else { /* argument doesn't exist */
        g_printerr("-display requires a parameter\n");
        help = TRUE;
      }
    } else if (0 == strcasecmp(argv[i], "-wmaker")) {
      wmaker = TRUE;
    } else if (0 == strcasecmp(argv[i], "-vertical")) {
      horizontal = FALSE;
    } else if (0 == strcasecmp(argv[i], "-border")) {
      ++i;

      if (i < argc) {
        int b = atoi(argv[i]);
        if (b > 0) {
          border = b;
        } else {
          g_printerr("-border must be a value greater than 0\n");
          help = TRUE;
        }
      } else { /* argument doesn't exist */
        g_printerr("-border requires a parameter\n");
        help = TRUE;
      }
      } else if (0 == strcasecmp(argv[i], "-iconsize")) {
      ++i;
      if (i < argc) {
        int s = atoi(argv[i]);
        if (s > 0) {
          icon_size = s;
        } else {
          g_printerr("-iconsize must be a value greater than 0\n");
          help = TRUE;
        }
      } else { /* argument doesn't exist */
        g_printerr("-iconsize requires a parameter\n");
        help = TRUE;
      }
    } else {
      if (argv[i][0] == '-')
        help = TRUE;
    }


    if (help) {
      /* show usage help */
      g_print("%s - version %s\n", argv[0], VERSION);
      g_print("Copyright 2003, Ben Jansens <ben@orodu.net>\n\n");
      g_print("Usage: %s [OPTIONS]\n\n", argv[0]);
      g_print("Options:\n");
      g_print("  -help             Show this help.\n");
      g_print("  -display DISLPAY  The X display to connect to.\n");
      g_print("  -border           The width of the border to put around the\n"
              "                    system tray icons. Defaults to 1.\n");
      g_print("  -vertical         Line up the icons vertically. Defaults to\n"
              "                    horizontally.\n");
      g_print("  -wmaker           WindowMaker mode. This makes docker a\n"
              "                    fixed size (64x64) to appear nicely in\n"
              "                    in WindowMaker.\n"
              "                    Note: In this mode, you have a fixed\n"
              "                    number of icons that docker can hold.\n");
      g_print("  -iconsize SIZE    The size (width and height) to display\n"
              "                    icons as in the system tray. Defaults to\n"
              "                    24.\n");
      exit(1);
    }
  }
}


void create_hint_win()
{
  XWMHints hints;
  XClassHint classhints;

  hint_win = XCreateSimpleWindow(display, root, 0, 0, 1, 1, 0, 0, 0);
  assert(hint_win);

  hints.flags = StateHint | WindowGroupHint | IconWindowHint;
  hints.initial_state = WithdrawnState;
  hints.window_group = hint_win;
  hints.icon_window = win;

  classhints.res_name = "docker";
  classhints.res_class = "Docker";

  XSetWMProperties(display, hint_win, NULL, NULL, argv, argc,
                   NULL, &hints, &classhints);

  XMapWindow(display, hint_win);
}


void create_main_window()
{
  XWMHints hints;
  XTextProperty text;

  /* the border must be > 0 if not in wmaker mode */
  assert(wmaker || border > 0);

  if (!wmaker)
    win = XCreateSimpleWindow(display, root, 0, 0,
                              border * 2, border * 2, 0, 0, 0);
  else
    win = XCreateSimpleWindow(display, root, 0, 0,
                              64, 64, 0, 0, 0);
    
  assert(win);

  char *name = "Docker";
  XStringListToTextProperty(&name, 1, &text);
  XSetWMName(display, win, &text);

  hints.flags = StateHint;
  hints.initial_state = WithdrawnState;
  XSetWMHints(display, win, &hints);
  
  create_hint_win();
  
  XSync(display, False);
  XSetWindowBackgroundPixmap(display, win, ParentRelative);
  XClearWindow(display, win);
}


void reposition_icons()
{
  int x = border + ((width % icon_size) / 2),
      y = border + ((height % icon_size) / 2);
  GSList *it;
  
  for (it = icons; it != NULL; it = g_slist_next(it)) {
    TrayWindow *traywin = it->data;
    traywin->x = x;
    traywin->y = y;
    XMoveWindow(display, traywin->id, x, y);
    XSync(display, False);
    if (wmaker) {
      x += icon_size;
      if (x + icon_size > width) {
        x = border;
        y += icon_size;
      }
    } else if (horizontal)
      x += icon_size;
    else
      y += icon_size;
  }
}


void fix_geometry()
{
  GSList *it;

  /* in wmaker mode we're a fixed size */
  if (wmaker) return;
  
  /* find the proper width and height */
  width = horizontal ? 0 : icon_size;
  height = horizontal ? icon_size : 0;
  for (it = icons; it != NULL; it = g_slist_next(it)) {
    TrayWindow *traywin = it->data;
    if (horizontal)
      width += icon_size;
    else
      height += icon_size;
  }

  XResizeWindow(display, win, width + border * 2, height + border * 2);
}


void event_loop()
{
  XEvent e;
  Window cover;
  GSList *it;
  
  while (!exit_app) {
    while (XPending(display)) {
      XNextEvent(display, &e);

      switch (e.type)
      {
      case PropertyNotify:
        /* systray window list has changed? */
        if (e.xproperty.atom == kde_systray_prop) {
          XSelectInput(display, win, NoEventMask);
          kde_update_icons();
          XSelectInput(display, win, StructureNotifyMask);

          while (XCheckTypedEvent(display, PropertyNotify, &e));
        }

        break;

      case ConfigureNotify:
        if (e.xany.window != win) {
          /* find the icon it pertains to and beat it into submission */
          GSList *it;
  
          for (it = icons; it != NULL; it = g_slist_next(it)) {
            TrayWindow *traywin = it->data;
            if (traywin->id == e.xany.window) {
              XMoveResizeWindow(display, traywin->id, traywin->x, traywin->y,
                                icon_size, icon_size);
              break;
            }
          }
          break;
        }
        
        /* briefly cover the entire containing window, which causes it and
           all of the icons to refresh their windows. finally, they update
           themselves when the background of the main window's parent changes.
        */
        cover = XCreateSimpleWindow(display, win, 0, 0,
                                    border * 2 + width, border * 2 + height,
                                    0, 0, 0);
        XMapWindow(display, cover);
        XDestroyWindow(display, cover);
        
        break;

      case ReparentNotify:
        if (e.xany.window == win) /* reparented to us */
          break;
      case UnmapNotify:
      case DestroyNotify:
        for (it = icons; it; it = g_slist_next(it)) {
          if (((TrayWindow*)it->data)->id == e.xany.window) {
            icon_remove(it);
            break;
          }
        }
        break;

      case ClientMessage:
        if (e.xclient.message_type == net_opcode_atom &&
            e.xclient.format == 32 &&
            e.xclient.window == net_sel_win)
          net_message(&e.xclient);
        
      default:
        break;
      }
    }
    usleep(500000);
  }

  /* remove/unparent all the icons */
  while (icons) {
    /* do the remove here explicitly, cuz the event handler isn't going to
       happen anymore. */
    icon_remove(icons);
  }
}

int main(int c, char **v)
{
  argc = c; argv = v;
  
  struct sigaction act;
  
  act.sa_handler = signal_handler;
  act.sa_flags = 0;
  sigaction(SIGSEGV, &act, NULL);
  sigaction(SIGPIPE, &act, NULL);
  sigaction(SIGFPE, &act, NULL);
  sigaction(SIGTERM, &act, NULL);
  sigaction(SIGINT, &act, NULL);
  sigaction(SIGHUP, &act, NULL);

  parse_cmd_line(argc, argv);

  display = XOpenDisplay(display_string);
  if (!display) {
    g_printerr("Unable to open Display %s. Exiting.\n",
               DisplayString(display_string));
  }

  root = RootWindow(display, DefaultScreen(display));
  assert(root);

  if (wmaker)
    width = height = 64 - border * 2;
  
  create_main_window();

  /* set up to find KDE systray icons, and get any that already exist */
  kde_init();

  net_init();

  /* we want to get ConfigureNotify events, and assume our parent's background
     has changed when we do, so we need to refresh ourself to match */
  XSelectInput(display, win, StructureNotifyMask);
  
  event_loop();

  XCloseDisplay(display);
}
